// react-stick/lib/utils/fit.js
"use strict";

exports.__esModule = true;
exports.fitsOnBottom = exports.fitsOnTop = exports.fitsOnLeft = exports.fitsOnRight = exports.isPositionedToLeft = exports.isPositionedToRight = exports.isPositionedToBottom = exports.isPositionedToTop = void 0;

var isPositionedToTop = function isPositionedToTop(position) {
  var _position$split = position.split(' '),
      positionMarker = _position$split[0];

  return positionMarker === 'top';
};

exports.isPositionedToTop = isPositionedToTop;

var isPositionedToBottom = function isPositionedToBottom(position) {
  var _position$split2 = position.split(' '),
      positionMarker = _position$split2[0];

  return positionMarker === 'bottom';
};

exports.isPositionedToBottom = isPositionedToBottom;

var isPositionedToRight = function isPositionedToRight(position) {
  var positionMarker = position.split(' ')[1];
  return positionMarker === 'right';
};

exports.isPositionedToRight = isPositionedToRight;

var isPositionedToLeft = function isPositionedToLeft(position) {
  var positionMarker = position.split(' ')[1];
  return positionMarker === 'left';
};

exports.isPositionedToLeft = isPositionedToLeft;

var fitsOnRight = function fitsOnRight(nodeRef, anchorRef) {
  var _nodeRef$getBoundingC = nodeRef.getBoundingClientRect(),
      nodeWidth = _nodeRef$getBoundingC.width;

  var _anchorRef$getBoundin = anchorRef.getBoundingClientRect(),
      anchorRight = _anchorRef$getBoundin.right;

  return anchorRight + nodeWidth <= window.innerWidth;
};

exports.fitsOnRight = fitsOnRight;

var fitsOnLeft = function fitsOnLeft(nodeRef, anchorRef) {
  var _nodeRef$getBoundingC2 = nodeRef.getBoundingClientRect(),
      nodeWidth = _nodeRef$getBoundingC2.width;

  var _anchorRef$getBoundin2 = anchorRef.getBoundingClientRect(),
      anchorLeft = _anchorRef$getBoundin2.left;

  return anchorLeft - nodeWidth >= 0;
};

exports.fitsOnLeft = fitsOnLeft;

var fitsOnTop = function fitsOnTop(nodeRef, anchorRef) {
  var _nodeRef$getBoundingC3 = nodeRef.getBoundingClientRect(),
      nodeHeight = _nodeRef$getBoundingC3.height;

  var _anchorRef$getBoundin3 = anchorRef.getBoundingClientRect(),
      anchorTop = _anchorRef$getBoundin3.top;

  return anchorTop - nodeHeight >= 0;
};

exports.fitsOnTop = fitsOnTop;

var fitsOnBottom = function fitsOnBottom(nodeRef, anchorRef) {
  var _nodeRef$getBoundingC4 = nodeRef.getBoundingClientRect(),
      nodeHeight = _nodeRef$getBoundingC4.height;

  var _anchorRef$getBoundin4 = anchorRef.getBoundingClientRect(),
      anchorBottom = _anchorRef$getBoundin4.bottom;

  return anchorBottom + nodeHeight <= window.innerHeight;
};

exports.fitsOnBottom = fitsOnBottom;
