// react-stick/lib/StickInline.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _substyle = _interopRequireDefault(require("substyle"));

var _utils = require("./utils");

function StickInline(_ref) {
  var node = _ref.node,
      children = _ref.children,
      component = _ref.component,
      containerRef = _ref.containerRef,
      nestingKey = _ref.nestingKey,
      align = _ref.align,
      position = _ref.position,
      style = _ref.style,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["node", "children", "component", "containerRef", "nestingKey", "align", "position", "style"]);
  var styles = (0, _substyle.default)(defaultStyle, {
    style: style
  }, (0, _utils.getModifiers)({
    align: align,
    position: position
  }));
  var Component = component || 'div';
  return /*#__PURE__*/_react.default.createElement(Component, (0, _extends2.default)({}, rest, styles, {
    ref: containerRef,
    "data-sticknestingkey": nestingKey
  }), children, node && /*#__PURE__*/_react.default.createElement("div", styles('node'), node));
}

var defaultStyle = {
  position: 'relative',
  node: {
    position: 'absolute',
    zIndex: 99,
    textAlign: 'left'
  },
  '&position-top': {
    node: {
      top: 0
    }
  },
  '&position-middle': {
    node: {
      top: '50%'
    }
  },
  '&position-bottom': {
    node: {
      top: '100%'
    }
  },
  '&position-left': {
    node: {
      left: 0
    }
  },
  '&position-center': {
    node: {
      left: '50%'
    }
  },
  '&position-right': {
    node: {
      left: '100%'
    }
  }
};
var _default = StickInline;
exports.default = _default;
module.exports = exports.default;
