// react-stick/lib/StickNode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _substyle = _interopRequireWildcard(require("substyle"));

var _utils = require("./utils");

function StickNode(_ref, ref) {
  var children = _ref.children,
      width = _ref.width,
      align = _ref.align,
      position = _ref.position,
      sameWidth = _ref.sameWidth;
  var styles = (0, _substyle.default)(defaultStyle, {}, (0, _utils.getModifiers)({
    align: align,
    position: position,
    sameWidth: sameWidth
  }));
  return /*#__PURE__*/_react.default.createElement("div", (0, _substyle.inline)(styles, {
    width: width
  }), /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, styles('content'), {
    ref: ref
  }), children));
}

var defaultStyle = {
  position: 'absolute',
  right: 0,
  bottom: 0,
  content: {
    // absolute position is needed as the stick node would otherwise
    // cover up the base node and, for instance, make it impossible to
    // click buttons
    position: 'absolute',
    display: 'inline-block',
    left: 'inherit',
    right: 'inherit',
    top: 'inherit',
    bottom: 'inherit'
  },
  '&sameWidth': {
    content: {
      display: 'block',
      width: '100%'
    }
  },
  '&align-left': {
    right: 'auto',
    left: 0
  },
  '&align-top': {
    bottom: 'auto',
    top: 0
  },
  '&align-middle': {
    content: {
      transform: 'translate(0, 50%)'
    }
  },
  '&align-center': {
    content: {
      transform: 'translate(50%, 0)'
    },
    '&align-middle': {
      content: {
        transform: 'translate(50%, 50%)'
      }
    }
  }
};

var _default = /*#__PURE__*/(0, _react.forwardRef)(StickNode);

exports.default = _default;
module.exports = exports.default;
